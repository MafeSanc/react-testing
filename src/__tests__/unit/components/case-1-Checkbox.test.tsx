import React from 'react'
import { axe } from 'jest-axe'
import { render, fireEvent } from '@testing-library/react'
import ReactDOM from 'react-dom'
import Checkbox from '../../../components/Checkbox'

/**
 * This checkbox component renders a checkbox with a label.
 * Since we customized the default checkbox, we want to
 * make sure it still works as a regular checkbox
 * should.
 */

describe('The <Checkbox /> component', () => {
  const defaultCheckboxProps = {
    label: 'TEST_CHECKBOX_LABEL',
    id: 'TEST_CHECKBOX_ID',
    background: '#000',
    checkMarkBackground: '#fff',
    onChange: jest.fn(),
    checked: false,
  }

  //this const is meant to avoid repetition
  const setupCheckbox = (props = defaultCheckboxProps) => render(<Checkbox {...props} />)

  it('❌ Should render the label and checkbox the user will see', () => {
    // const container = document.createElement('div')
    // ReactDOM.render(
    //   <Checkbox {...defaultCheckboxProps} />,
    //   container
    // )

    // expect(container.querySelector('label')).not.toBeNull()
    // expect(container.querySelector('input[type="checkbox"]')).not.toBeNull()

    // const{ getByLabelText, debug} = render(
    //   <Checkbox {...defaultCheckboxProps}  />,
    // )
    // debug(getByLabelText('TEST_CHECKBOX_LABEL'))

    const { asFragment } = render(<Checkbox {...defaultCheckboxProps} />)
    // console.log(asFragment())
    expect(asFragment()).toMatchInlineSnapshot()
  })

  it(' Should make the checkbox accessible by setting the id and htmlFor attributes on label and checkbox', () => {
    const { getByLabelText, debug } = render(
      <Checkbox {...defaultCheckboxProps} />,
    )

    expect(getByLabelText('TEST_CHECKBOX_LABEL')).toBeInTheDocument()
  })


  it('❌ Should call the onChange handler when it is provided', () => {
    const { getByLabelText } = setupCheckbox()

    const checkbox = getByLabelText(defaultCheckboxProps.label)

    fireEvent.click(checkbox)

    expect(defaultCheckboxProps.onChange).toHaveBeenCalled()
  })


  it('❌ Should change state correctly when clicked (checked and unchecked)', () => {
    const { getByLabelText } = setupCheckbox({
      ...defaultCheckboxProps,
      checked: true
    })

    expect(getByLabelText(defaultCheckboxProps.label)).toBeChecked()

  })

  it('❌ should not fail any accessibility tests', async () => {})
})
