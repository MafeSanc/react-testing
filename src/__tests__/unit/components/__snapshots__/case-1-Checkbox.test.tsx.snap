// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The <Checkbox /> component  Should make the checkbox accessible by setting the id and htmlFor attributes on label and checkbox 1`] = `
<DocumentFragment>
  <div
    class="sc-AxjAm StDqN"
    data-testid="Checkbox"
  >
    <input
      class="sc-AxhCb iqvdaa"
      id="TEST_CHECKBOX_ID"
      type="checkbox"
    />
    <label
      class="sc-AxiKw cwRuvq"
      for="TEST_CHECKBOX_ID"
    >
      <span
        class="sc-AxirZ bZbnmP"
      >
        TEST_CHECKBOX_LABEL
      </span>
    </label>
  </div>
</DocumentFragment>
`;
